prog/mutation.go

```diff
@@ -419,49 +466,128 @@ type mutationArgs struct {
 			return
 		}
 	}
+
+	prio, flagStop, canMutate := arg.Type().getMutationPrio()
+	ctx.Stop = flagStop
+
+	if !canMutate {
+		return
+	}
+
 	typ := arg.Type()
 	if typ == nil || typ.Dir() == DirOut || !typ.Varlen() && typ.Size() == 0 {
 		return
 	}
+
 	ma.args = append(ma.args, arg)
 	ma.ctxes = append(ma.ctxes, *ctx)
+	ma.priorities = append(ma.priorities, prio)
+}
+
+// TODO: find a way to estimate optimal priority values.
+// Assign a priority for each type. The boolean is the reference type and it has
+// the minimum priority, since it has only two possible values.
+func (t *IntType) getMutationPrio() (prio float64, ctxStop bool, canMutate bool) {
+	if (t.RangeEnd - t.RangeBegin) == 1 {
+		return minPriority, true, true
+	}
+
+	bitsCount := math.Log2((float64(t.TypeSize * 8)))
+	if t.RangeBegin == t.RangeEnd {
+		return bitsCount + 2, true, true
+	}
+
+	// an integer with a range could have a higher impact
+	return bitsCount + 4, true, true
+}
+
+func (t *StructType) getMutationPrio() (prio float64, ctxStop bool, canMutate bool) {
+	return maxPriority, true, true
+}
+
+func (t *UnionType) getMutationPrio() (prio float64, ctxStop bool, canMutate bool) {
+	if len(t.Fields) == 1 {
+		return minPriority, false, false
+	}
+
+	return maxPriority, true, true
+}
+
+func (t *FlagsType) getMutationPrio() (prio float64, ctxStop bool, canMutate bool) {
+	if len(t.Vals) == 0 { // No value to mutate
+		return minPriority, true, false
+	}
+
+	// when bitmask is true we should a have a higher priority because we can have bitwise operations
+	if t.BitMask {
+		bitsCount := math.Log2((float64(t.TypeSize * 8)))
+		return bitsCount + 3, true, true // a higher priority than an integer without range
+	}
+
+	return 0.6 * maxPriority, true, true
+}
+
+func (t *PtrType) getMutationPrio() (prio float64, ctxStop bool, canMutate bool) {
+	return 3, false, true
+}
+
+func (t *ConstType) getMutationPrio() (prio float64, ctxStop bool, canMutate bool) {
+	return minPriority, true, false
+}
+
+func (t *CsumType) getMutationPrio() (prio float64, ctxStop bool, canMutate bool) {
+	return minPriority, true, false
+}
+
+func (t *ProcType) getMutationPrio() (prio float64, ctxStop bool, canMutate bool) {
+	return 0.5 * maxPriority, true, true
+}
+
+func (t *ResourceType) getMutationPrio() (prio float64, ctxStop bool, canMutate bool) {
+	return 0.5 * maxPriority, true, true
+}
+
+func (t *VmaType) getMutationPrio() (prio float64, ctxStop bool, canMutate bool) {
+	return 0.5 * maxPriority, true, true
+}
+
+func (t *LenType) getMutationPrio() (prio float64, ctxStop bool, canMutate bool) {
+	return 0.6 * maxPriority, true, true
+}
+
+func (t *BufferType) getMutationPrio() (prio float64, ctxStop bool, canMutate bool) {
+	return 0.8 * maxPriority, true, true
+}
+
+func (t *ArrayType) getMutationPrio() (prio float64, ctxStop bool, canMutate bool) {
+	if t.Kind == ArrayRangeLen && t.RangeBegin == t.RangeEnd {
+		return minPriority, true, false
+	}
+	return maxPriority, true, true
```

Comment:
What test covers this?
I see a new `mutate$array(0x1, 0x30, &(0x7f0000000000)=[0x1, 0x1, 0x1, 0x1, 0x1])` test, but it tests mutation of the array itself (changing number of elements) rather than mutation of the elements of the array. Did it fail with the previous version of code? We returned canMutate=true for arrays, so it should have been passed unless I am missing something.